nomad/worker.go

```diff
@@ -70,70 +113,286 @@ type Worker struct {
+// status of the worker based on calls to the Worker API.
+func (w *Worker) setStatus(newStatus WorkerStatus) {
+	w.statusLock.Lock()
+	defer w.statusLock.Unlock()
+	if newStatus == w.status {
+		return
+	}
+	w.logger.Trace("changed worker status", "from", w.status, "to", newStatus)
+	w.status = newStatus
+}
+
+// GetStatus returns the status of the Worker's Workload.
+func (w *Worker) GetWorkloadStatus() SchedulerWorkerStatus {
+	w.statusLock.RLock()
+	defer w.statusLock.RUnlock()
+	return w.workloadStatus
+}
+
+// setWorkloadStatus is used internally to the worker to update the
+// status of the worker based updates from the workload.
+func (w *Worker) setWorkloadStatus(newStatus SchedulerWorkerStatus) {
+	w.statusLock.Lock()
+	defer w.statusLock.Unlock()
+	if newStatus == w.workloadStatus {
+		return
+	}
+	w.logger.Trace("changed workload status", "from", w.workloadStatus, "to", newStatus)
+	w.workloadStatus = newStatus
+}
+
+type WorkerInfo struct {
+	ID                string    `json:"id"`
+	EnabledSchedulers []string  `json:"enabled_schedulers"`
+	Started           time.Time `json:"started"`
+	Status            string    `json:"status"`
+	WorkloadStatus    string    `json:"workload_status"`
+}
+
+func (w WorkerInfo) Copy() WorkerInfo {
+	out := WorkerInfo{
+		ID:                w.ID,
+		EnabledSchedulers: make([]string, len(w.EnabledSchedulers)),
+		Started:           w.Started,
+		Status:            w.Status,
+		WorkloadStatus:    w.WorkloadStatus,
+	}
+	copy(out.EnabledSchedulers, w.EnabledSchedulers)
+	return out
+}
+
+func (w WorkerInfo) String() string {
+	// lazy implementation of WorkerInfo to string
+	out, _ := json.Marshal(w)
+	return string(out)
+}
+
+func (w *Worker) Info() WorkerInfo {
+	w.pauseLock.Lock()
+	defer w.pauseLock.Unlock()
+	out := WorkerInfo{
+		ID:                w.id,
+		Status:            w.status.String(),
+		WorkloadStatus:    w.workloadStatus.String(),
+		EnabledSchedulers: make([]string, len(w.enabledSchedulers)),
+	}
+	out.Started = w.start
+	copy(out.EnabledSchedulers, w.enabledSchedulers)
+	return out
+}
+
+// ----------------------------------
+//  Pause Implementation
+//    These functions are used to support the worker's pause behaviors.
+// ----------------------------------
+
+func (w *Worker) setPauseFlag(pause bool) {
+	w.pauseLock.Lock()
+	defer w.pauseLock.Unlock()
+	w.pauseFlag = pause
+}
+
+// maybeWait is responsible for making the transition from `pausing`
+// to `paused`, waiting, and then transitioning back to the running
+// values.
+func (w *Worker) maybeWait() {
 	w.pauseLock.Lock()
-	for w.paused {
+	if !w.pauseFlag {
+		w.pauseLock.Unlock()
+		return
+	}
+
+	w.statusLock.Lock()
+	w.status = WorkerPaused
+	originalWorkloadStatus := w.workloadStatus
+	w.workloadStatus = WorkloadPaused
+	w.logger.Trace("changed workload status", "from", originalWorkloadStatus, "to", w.workloadStatus)
+
+	w.statusLock.Unlock()
+
+	for w.pauseFlag {
 		w.pauseCond.Wait()
 	}
+
+	w.pauseLock.Unlock()
```

Comment:
I think there's a race here if a `Pause` method call comes in after this line but before the status is set before we return. Even with the locks the `Pause` method's status setting calls could be interleaved such that the worker sets its status to `WorkerPaused`, sets the pause flag, and then its status is set to `WorkerStarted`. 

Maybe we should move this up to the top of the function as a `defer w.pauseLock.Unlock()`?